### PR TITLE
Fully encapsulate LocalCheckpointTracker inside of the engine

### PR DESCRIPTION
**Pull request by dakrone**
This makes the Engine interface not expose the `LocalCheckpointTracker`, instead
exposing the pieces needed (like retrieving the local checkpoint) as individual
methods.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 1]
```diff
@@ -635,11 +635,38 @@ public CommitStats commitStats() {
...
     /**
-     * The sequence number service for this engine.
+     * @return the local checkpoint for this Engine
+     */
+    public abstract long getLocalCheckpoint();
+
+    /**
+     * @return the maximum sequence number from the local checkpoint tracker
+     */
+    public abstract long getMaxSeqNo();
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/Engine.java]: maybe we can remove this? it's only used in one place where I think we can just use getSeqNoStats?

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 2]
```diff
@@ -635,11 +635,38 @@ public CommitStats commitStats() {
...
+
+    /**
+     * @return generate a new sequence number from the local checkpoint tracker
+     */
+    public abstract long generateSeqNo();
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/Engine.java]: this is super dangerous IMO. Maybe we can make `getLocalCheckpointTracker` package private and have a testing utility to get it? I think that's good enough for tests and we can keep it isolated from production code?

**Review by bleskes — CHANGES_REQUESTED** [review]
Thank you @dakrone ! I left one comment about generateSeqNo that I think is important. The rest is just a suggestion.

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java (hunk 1 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/Engine.java]: Certainly, I've done that, passing in -1 for the globalCheckpoint

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java (hunk 2 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/Engine.java]: Yep, I've removed this

**Review by bleskes — APPROVED** [review]
LGTM. Thanks @dakrone

**Pull request merged**